### DISCUSSION TITLE
EventToCommandBehavior Enable/Disable

### DISCUSSION THREAD

**Discussion by FM1973** [CommunityToolkit/Maui#724] [General]
Hello,
It would be nice if one could enable and disable an EventToCommandBehavior.

In my case I use the EventToCommandBehavior for Binding a command if a switch in a collectionview is toggled.
this works pretty well. The problem is, the EventToCommandBehavior also calls the command during loading the collectionview items.

**Comment by bijington** [comment]
You can control whether the Command will be executed using the CanExecute parameter. I appreciate this would rely on having some knowledge in the View Model that owns the Command property. Would that work?

**Reply by FM1973** [reply]
Thanks for your suggestion.
The problem is the same.
Loading data finishes in the viewmodel and afterwards the content get´s rendered.
In that case the "LoadedProperty" is already true before the page gets rendered and the command gets executed.
I solved this, in my case, by checking if the list items really got updated and only then save them.
But the root problem remains the same.
If there was a "Rendered-Event" of the collectionview I could get around that. But there is none.
So even a "Enabled" Property of the EventToCommandBehavior would´t help. Because this would be set too early as well.

**Reply by bijington** [reply]
Oh yes the Loaded event is likely not what you need. Can we take a step back and you explain what it is that you are trying to achieve?

**Reply by FM1973** [reply]
Sure.

I got a list of opening times for a store. The list is being displayed as a collectionview.

Each entry in the list has a "enabled" property (means the opening time should be displayed for other users or not).
For every entry in the list there is a switch. When the user toggles the switch (enable/disable opening time), I save the new state of the entry to the database (and on the server via API). So there is no "general" Save button.

Since there is no Command.Property available for a switch, I use the EventToCommandBehavior to call the "Save Command" when a user toggles a switch in the list.

The problem now is that when I load the list and bind it using an observable collection, the Toggled-Event of the switch is being called during rendering the collectionview and my "save-code" is being called although there has been no user interaction.

Normally I use some sort of "IsBusy" property which is set to true before loading the list and is set to false once the list has been loaded and bound. But the problem using this property is, that it is being set to false before the list gets rendered. So I cant´s use that property.

I hope this info helps understanding the problem.

**Reply by IeuanWalker** [reply]
@FM1973 I believe that is normal behavior even if you use the actual event.

Are you able to compare the old value to the new value, and only do the API calls if they differ?

**Reply by FM1973** [reply]
How can I do that?
Currently I´m using the items ID as a parameter for the command.
A RelayCommand can have only on parameter.
I tried to change the command parameter binding to the item itself (entity) using "{Binding .}" but the parameter then is always null. Which is strange, because when I bind to ID "{Binding Id}" it works as expected.
But I think even if I could get the entity itself, the value of the enabled property would have already changed and the comparison would´t make sense.

**Reply by bijington** [reply]
If you have a view model that represents each item in the list then you could potentially have it handle the saving and you trigger that based on when the Switch is toggled through a binding to a boolean property on the view model. If that makes sense?

**Reply by IeuanWalker** [reply]
@FM1973 you could also have 2 bools in your model.

i.e. IntialValue + CurrentValue
Binding the CurrentValue to Toggle, when the command is triggered compare the `IntialValue == CurrentValue` and then do the API call if they differ.
If the API call is successful, update the `IntialValue = CurrentValue`

**Reply by IeuanWalker** [reply]
And as @bijington you could also handle it in the bindings and not have an event or command.

Using the new [MVVM toolkit](https://learn.microsoft.com/en-us/dotnet/communitytoolkit/mvvm/generators/observableproperty) you can use the `OnChanging` method, which will give you the value it's about to change too. So you can compare the parameter value to the model value, i.e. -
![image](https://user-images.githubusercontent.com/6544051/199752825-1a5be7b0-0160-4ab4-be06-96eada7f0fe1.png)

**Reply by FM1973** [reply]
Thanks for your suggestions.
I´m talking about a collectionview of items (ObservableCollection). So OnXXXChanging won´t work.

**Reply by bijington** [reply]
it can work if the items in your `ObservableCollection` are observable objects. A brief example is below but note I have written it in the browser so it may not be complete.

```csharp
public class ItemViewModel
{
    // Handle OnXXXChanging in here
}

public class MainViewModel
{
    public ObservableCollection<ItemViewModel> Items { get; }
}
```

**Comment by CliffAgius** [comment]
Closed and answered
